include/SharedMemory.h

```diff
@@ -0,0 +1,143 @@
+/*
+ * SharedMemory.h
+ *
+ * Copyright (c) 2022 Dominic Clark <mrdomclark/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ */
+
+#ifndef SHARED_MEMORY_H
+#define SHARED_MEMORY_H
+
+#include <memory>
+#include <string>
+#include <type_traits>
+
+namespace detail {
+
+class SharedMemoryImpl;
+
+class SharedMemoryData
+{
+public:
+	SharedMemoryData() noexcept;
+	SharedMemoryData(std::string&& key, bool readOnly);
+	SharedMemoryData(std::string&& key, std::size_t size, bool readOnly);
+	~SharedMemoryData();
+
+	SharedMemoryData(SharedMemoryData&& other) noexcept;
+	SharedMemoryData& operator=(SharedMemoryData&& other) noexcept
+	{
+		auto temp = std::move(other);
+		swap(*this, temp);
+		return *this;
+	}
+
+	friend void swap(SharedMemoryData& a, SharedMemoryData& b) noexcept
+	{
+		using std::swap;
+		swap(a.m_key, b.m_key);
+		swap(a.m_impl, b.m_impl);
+		swap(a.m_ptr, b.m_ptr);
+	}
+
+	const std::string& key() const noexcept { return m_key; }
+	void* get() const noexcept { return m_ptr; }
+
+private:
```

Comment:
> It would.
> 
> [class.base.init]/9:
> 
> > In a non-delegating constructor, if a given potentially constructed subobject is not designated by a _mem-initializer-id_ [...] the entity is default-initialized.
> 
> [dcl.init]/7:
> 
> > To _default-initialize_ an object of type `T` means:
> > 
> > * If `T` is a (possibly cv-qualified) class type, constructors are considered. The applicable constructors are enumerated ([over.match.ctor]), and the best one for the _initializer_ `()` is chosen through overload resolution. The constructor thus selected is called, with an empty argument list, to initialize the object.
> > * If `T` is an array type, each element is default-initialized.
> > * Otherwise, no initialization is performed.
> 
> `std::string` and `std::unique_ptr` have default constructors that initialise them to the empty string and `nullptr` respectively, which is the desired behaviour here (and indeed equivalent to your suggestion). On the other hand, `void*` is not a class type (or an array type), so no initialisation is performed by default-initialization and thus the explicit initialisation is required.

Ah, okay. I usually strive for just a consistent look across the board, so initalizing only ``m_ptr`` seemed unusal to me at first.